### PR TITLE
Implement alarm raise and clear

### PR DESCRIPTION
**Pull request by dpino**
This PR implements the alarm raise and clear logic.

An alarm is raised on calling `alarm:raise()` in `lib.yang.alarms`. An alarm keeps a track of status changes. There are 3 conditions on which a new status change is created for an alarm:
- `is_cleared` changes.
- `severity_priority` changes.
- `alarm_text` changes.

The field `is_cleared` marks whether an alarm was cleared by the system or not. When an alarm is raised, `is_cleared` is false, while on running `alarm:clear` this field is set to true.

The alarm:raise method admits a list or arguments that can set the fields of an alarm. For instance, an alarm is raise with priority 'minor'.

```lua
alarm:raise({perceived_severity='minor'})
```

After several attempts, the priority of the alarm increases to critical:

```lua
alarm:raise({perceived_severity='critical'})
```

The change is priority will cause a new status change in the alarm.

### CODE REVIEW

**Comment by dpino** [comment]
Example of use.

**lwaftr**

```bash
$ sudo ./snabb lwaftr run --name lwaftr --conf icmp_on_fail.conf --on-a-stick 83:00.0
```

**get-state**

```bash
$ sudo ./snabb config get-state lwaftr /softwire-state/alarms                                                                                                  
alarm-inventory {
  alarm-type {
    alarm-type-id arp-resolution;
    alarm-type-qualifier ;
    description "Raise up if ARP app cannot resolve IP address";
    has-clear true;
    resource external-interface;
  }
  alarm-type {
    alarm-type-id ndp-resolution;
    alarm-type-qualifier ;
    description "Raise up if NDP app cannot resolve IPv6 address";
    has-clear true;
    resource internal-interface;
  }
}
alarm-list {
  alarm {
    alarm-type-id arp-resolution;
    alarm-type-qualifier ;
    resource external-interface;
    alarm-text "Make sure you can resolve external-interface.next-hop.ip address manually.  If it cannot be resolved, consider setting the MAC address of the next-hop directly.  To do it so, set external-interface.next-hop.mac to the value of the MAC address.";
    is-cleared false;
    last-changed 2017-08-28T11:00:56Z;
    perceived-severity critical;
    status-change {
      time 2017-08-28T11:00:56Z;
      alarm-text "Make sure you can resolve external-interface.next-hop.ip address manually.  If it cannot be resolved, consider setting the MAC address of the next-hop directly.  To do it so, set external-interface.next-hop.mac to the value of the MAC address.";
      perceived-severity critical;
    }
    time-created 2017-08-28T11:00:56Z;
  }
  last-changed 2017-08-28T11:00:56Z;
  number-of-alarms 1;
}
```

File: src/lib/yang/alarms.lua [hunk 1]
```diff
@@ -84,6 +109,20 @@ end
 
 -- Contains a table with all the declared alarms.
 local alarm_list = {}
+function alarm_list:lookup (key)
+   return alarm_list[key]
```
Comment (wingo) [inline src/lib/yang/alarms.lua]: Since `lookup` is also a member of `alarm_list`, perhaps you would prefer to separate things a bit; store the list in a member of the `alarm_list` table?

File: src/lib/yang/util.lua [hunk 2]
```diff
@@ -129,6 +129,20 @@ function memoize(f, max_occupancy)
...
+local function gmtime ()
+   local now = os.time()
+   local utcdate = os.date("!*t", now)
+   local localdate = os.date("*t", now)
+   localdate.isdst = false
+   local timediff = os.difftime(os.time(utcdate), os.time(localdate))
+   return now + timediff
+end
+
+function iso_8601 (time)
```
Comment (wingo) [inline src/lib/yang/util.lua]: Can we rename this function?  `format_date_as_iso_8601` maybe -- too long but at least it includes the word `date`.  Dunno, other ideas possible, otherwise it's a bit hard to read.

File: src/lib/yang/alarms.lua [hunk 3]
```diff
@@ -101,26 +140,207 @@ function declare_alarm (alarm)
...
+   alarm.perceived_severity = status.perceived_severity
+   alarm.alarm_text = status.alarm_text
+   alarm.last_changed = status.time
+   state.alarm_list.last_changed = status.time
+   table.insert(alarm.status_change, status)
+end
+
+-- Creates a new alarm.
+--
+-- The alarm is retrieved from the db of predefined alarms. Default values got
```
Comment (wingo) [inline src/lib/yang/alarms.lua]: "get" overridden I think

File: src/lib/yang/alarms.lua [hunk 4]
```diff
@@ -101,26 +140,207 @@ function declare_alarm (alarm)
...
+      alarm.is_cleared = args.is_cleared
+   end
+end
+
+-- Check up if the alarm already exists in state.alarm_list.
+local function lookup_alarm (key)
+   return state.alarm_list.alarm[key]
+end
+
+-- Codec serializes nil values as empty strings. Convert them back to nil values.
```
Comment (wingo) [inline src/lib/yang/alarms.lua]: Why does the codec do this?  You can't detect the presence or absence of a nil-valued property anyway.

**Review by wingo — COMMENTED** [review]
Looking good, some small comments and questions.

File: src/lib/yang/alarms.lua (hunk 4 above)
Comment (dpino) [reply src/lib/yang/alarms.lua]: Some arguments in `alarm:raise` might be nil. For instance:

```lua
alarm:raise(key, {perceived_severity='minor'})
```

In this call `alarm_text` is nil.

`codec:string` cannot encode a `nil` value, so I have to encode `nil` arguments as something (an empty string). If I don't do it and let `finish()` to write only set values, it won't be possible to distinguish which argument is which.

Since all arguments have a value (even when they were originally `nil`), I need to do this conversion back to nil. If I don't do it, in the example above `alarm_text` will be empty and the description text of the alarm will get overwritten to an empty string.

Comment (wingo) [reply src/lib/yang/alarms.lua]: Could you make the codec define a `maybe_string` type?  Like if it writes UINT32_MAX for the length, then it's nil, otherwise it's a string.  dunno.  E.g.

```lua
local UINT32_MAX = 0xffffffff
function encoder:maybe_string(str)
   if str == nil then
      self:uint32(UINT32_MAX)
   else
      self:string(str)
   end
end

...

   function decoder:maybe_string(str)
      local len = self:uint32()
      if len == UINT32_MAX then return nil end
      return ffi.string(self:read(len), len)
   end
```

then use maybe_string wherever you would want to encode/decode a value that may be a string or nil.  That way you can actually encode the empty string if you need to!

File: src/lib/yang/alarms.lua [hunk 5]
```diff
@@ -83,14 +108,33 @@ local function table_size (t)
...
 -- Contains a table with all the declared alarms.
-local alarm_list = {}
+local alarm_list = {
+   list = {},
+}
+function alarm_list:new (key, alarm)
+   self.list[key] = alarm
+end
+function alarm_list:lookup (key)
+   return alarm_list.list[key]
```
Comment (wingo) [inline src/lib/yang/alarms.lua]: should be `self.list` perhaps?

File: src/lib/yang/alarms.lua [hunk 6]
```diff
@@ -83,14 +108,33 @@ local function table_size (t)
...
+end
+function alarm_list:copy (src, args)
```
Comment (wingo) [inline src/lib/yang/alarms.lua]: Perhaps should be a private function?

File: src/apps/config/alarm_codec.lua [hunk 7]
```diff
@@ -120,7 +134,7 @@ local function normalize (t, attrs)
    t = t or {}
    local ret = {}
    for _, k in ipairs(attrs) do
-      table.insert(ret, t[k] or '')
+      table.insert(ret, t[k])
```
Comment (wingo) [inline src/apps/config/alarm_codec.lua]: Does this do the right thing if `t[k]` is nil?  Would seem to me like `#ret` would stay the same.

**Review by wingo — APPROVED** [review]
Looks good, a few questions to fix either in this or in a followup.

File: src/apps/config/alarm_codec.lua (hunk 7 above)
Comment (dpino) [reply src/apps/config/alarm_codec.lua]: The goal of this function is to decompress a sparse table into an list of values in a specific order. For instance in the call:

```lua
alarm:raise(key, {alarm_text='text'})
```

The function decompresses `args` ({alarm_text='text'}) to arguments `perceived_severity` and `alarm_text` (which are stored in ret). In this case `args` doesn't define `perceived_severity` so `ret` will contain {nil, 'text'}. The size of `ret` is always the same regardless of the size of #t (sometimes might be the same, but decompressing the list in a specific order is important too).

Comment (wingo) [reply src/apps/config/alarm_codec.lua]:
```
wingo@sparrow ~/src/snabb/src$ cat /tmp/foo.lua
local x = {}
print(#x)
table.insert(x, nil)
print(#x)
wingo@sparrow ~/src/snabb/src$ sudo ./snabb snsh /tmp/foo.lua
0
0
```

Are you sure it would end up being `{nil, 'text'}` ?  Looks to me like it might be `{'text'}`.

Comment (dpino) [reply src/apps/config/alarm_codec.lua]: Yep, you're right. It doesn't work as I expected. Inserting a nil in a table has no effect.

If I leave an empty string as a  placeholder for nil values, all arguments will be encoded as strings. OTOH, I can use something else as a placeholder and make `maybe_string` to consider that placeholder a nil value. WDYT?

Comment (wingo) [reply src/apps/config/alarm_codec.lua]: If your intention is really to unroll some possibly-nil positional arguments, then perhaps use:
```
local args = {}
for i,k in ipairs(attrs) do args[i] = t[k] done
```

File: src/apps/config/alarm_codec.lua [hunk 8]
```diff
@@ -184,7 +196,7 @@ function selftest ()
       assert(lib.equal({name, args}, decoded))
    end
    local function test_raise_alarm ()
-      local key = {resource='res1'}
+      local key = {resource='res1', alarm_type_id='type1', alarm_type_qualifier=''}
```
Comment (wingo) [inline src/apps/config/alarm_codec.lua:199]: Possibly revert this change so that we are testing the right thing?  And below.

Comment (dpino) [reply src/apps/config/alarm_codec.lua:199]: I left `resource`, `alarm_type_id` and `alarm_type_qualifier` encoded as strings, because according to the YANG schema they are mandatory fields (I considered though whether to give `alarm_type_qualifier` a default empty string value in case of `nil`, because I believe many times this value won't be used).

The `maybe_string` functionality is tested via `local args`. In this case, all arguments in `args` might be nil. The test only uses `perceived_severity` and it's working fine, so that means `maybe_string` is tested and working OK.
